File: app/src/main/java/eu/warble/pjapp/data/remote/PjatkAPI.java
package eu.warble.pjapp.data.remote;

import android.util.Log;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import eu.warble.pjapp.data.StudentDataSource;
import eu.warble.pjapp.data.model.Student;
import eu.warble.pjapp.util.AppExecutors;
import eu.warble.pjapp.util.Constants;
import eu.warble.pjapp.util.Converter;
import eu.warble.pjapp.util.CredentialsManager.Credentials;
import eu.warble.pjapp.util.NTLMAuthenticator;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class PjatkAPI implements StudentDataSource{

    private static PjatkAPI INSTANCE;
    private final OkHttpClient httpClient;
    private final AppExecutors appExecutors;
    private static final String STUDENT_PERSONAL_DATA_JSON;
    private static final String STUDENT_SCHEDULE;
    private static final String CHECK_CREDENTIALS;
    private static final int WRONG_CREDENTIALS = 401;
    private static final int OK = 200;

    static {
        STUDENT_PERSONAL_DATA_JSON = "https://ws.pjwstk.edu.pl/test/Service.svc/XmlService/GetStudentPersonalDataJson";
        STUDENT_SCHEDULE = "https://ws.pjwstk.edu.pl/test/Service.svc/XmlService/GetStudentSchedule";
        CHECK_CREDENTIALS = "https://ws.pjwstk.edu.pl/test/Service.svc/XmlService/tester";
    }

    // Prevent direct instantiation.
    private PjatkAPI(Credentials credentials, AppExecutors appExecutors){
        this.appExecutors = appExecutors;
        httpClient = new OkHttpClient.Builder()
                .authenticator(new NTLMAuthenticator(credentials.login, credentials.password))
                .connectTimeout(8, TimeUnit.SECONDS)
                .build();
    }

    public static PjatkAPI getInstance(Credentials credentials, AppExecutors appExecutors) {
        if (INSTANCE == null) {
            INSTANCE = new PjatkAPI(credentials, appExecutors);
        }
        return INSTANCE;
    }

    public static void destroyInstance(){
        INSTANCE = null;
    }

    public void getStudentData(LoadStudentDataCallback callback){
        if (httpClient == null){
            Log.e("PjatkApi", "httpClient is not initialized");
            callback.onDataNotAvailable(Constants.UNKNOWN_ERROR);
            return;
        }

        Request request = new Request.Builder().url(STUDENT_PERSONAL_DATA_JSON).build();
        appExecutors.networkIO().execute(() -> {
            try (Response response = httpClient.newCall(request).execute()){
                switch (response.code()) {
                    case WRONG_CREDENTIALS:
                        appExecutors.mainThread().execute(() -> callback.onDataNotAvailable(Constants.CREDENTIALS_ERROR));
                        break;
                    case OK:
                        ResponseBody body = response.body();
                        if (body == null)
                            appExecutors.mainThread().execute(() -> callback.onDataNotAvailable(Constants.CONNECTION_ERROR));
                        else {
                            Student studentData = Converter.jsonStringToStudentData(Converter.responseToJsonString(body));
                            appExecutors.mainThread().execute(() -> callback.onDataLoaded(studentData));
                        }
                        break;
                    default:
                        appExecutors.mainThread().execute(() -> callback.onDataNotAvailable(Constants.UNKNOWN_ERROR));
                        Log.e("PjatkApi", "Wrong response from PJATK server, response : " + response.toString());
                        break;
                }
            }catch (IOException ex){
                appExecutors.mainThread().execute(() -> callback.onDataNotAvailable(Constants.CONNECTION_ERROR));
            }
        });
    }
}

File: app/src/main/java/eu/warble/pjappkotlin/data/remote/PjatkAPI.kt
package eu.warble.pjappkotlin.data.remote

import android.content.Context
import android.util.Log
import eu.warble.pjappkotlin.data.ScheduleDataSource
import eu.warble.pjappkotlin.utils.Constants
import eu.warble.pjappkotlin.data.StudentDataSource
import eu.warble.pjappkotlin.utils.AppExecutors
import eu.warble.pjappkotlin.utils.Converter
import eu.warble.pjappkotlin.utils.CredentialsManager
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.threeten.bp.LocalDate
import java.util.concurrent.TimeUnit

class PjatkAPI private constructor(credentials: CredentialsManager.Credentials) : StudentDataSource, ScheduleDataSource {

    private var httpClient: OkHttpClient? = null

    init {
        val login = credentials.login
        val password = credentials.password
        if (login != null && password != null) {
            httpClient = OkHttpClient.Builder()
                    .authenticator(NTLMAuthenticator(login, password))
                    .connectTimeout(8, TimeUnit.SECONDS)
                    .build()
        }
    }

    override fun getStudentData(appContext: Context, callback: StudentDataSource.LoadStudentDataCallback) {
        AppExecutors.NETWORK().execute {
            makeRequest(Constants.API_STUDENT_PERSONAL_DATA_JSON, object : ResponseCallback {
                override fun onResponseLoaded(response: Response) {
                    val body = response.body()
                    if (body != null) {
                        val studentData = Converter.jsonStringToStudentData(
                                Converter.responseToString(body)
                        )
                        if (studentData != null) {
                            AppExecutors.MAIN().execute { callback.onDataLoaded(studentData) }
                        } else {
                            onError(Constants.UNKNOWN_ERROR)
                        }
                    } else {
                        onError(Constants.UNKNOWN_ERROR)
                    }
                }

                override fun onError(error: String) {
                    AppExecutors.MAIN().execute { callback.onDataNotAvailable(error) }
                }
            })
        }
    }

    override fun getScheduleData(appContext: Context, from: LocalDate, to: LocalDate, callback: ScheduleDataSource.LoadScheduleDataCallback) {
        AppExecutors.NETWORK().execute {
            makeRequest("${Constants.API_STUDENT_SCHEDULE}&${from.format(Constants.API_DATE_FORMAT)}&${to.format(Constants.API_DATE_FORMAT)}",
                    object : ResponseCallback {
                        override fun onResponseLoaded(response: Response) {
                            val body = response.body()
                            if (body != null) {
                                val scheduleData = Converter.jsonStringToScheduleList(
                                        Converter.responseToString(body)
                                )
                                if (scheduleData != null) {
                                    AppExecutors.MAIN().execute { callback.onDataLoaded(scheduleData) }
                                } else {
                                    onError(Constants.UNKNOWN_ERROR)
                                }
                            } else {
                                onError(Constants.UNKNOWN_ERROR)
                            }
                        }

                        override fun onError(error: String) {
                            AppExecutors.MAIN().execute { callback.onDataNotAvailable(error) }
                        }
                    }
            )
        }
    }

    private fun makeRequest(url: String, responseCallback: ResponseCallback) {
        if (httpClient == null) {
            Log.e("PjatkApi", "httpClient is not initialized")
            responseCallback.onError(Constants.UNKNOWN_ERROR)
            return
        }
        val request = Request.Builder().url(url).build()
        try {
            httpClient?.newCall(request)?.execute().use {
                when (it?.code()) {
                    OK -> responseCallback.onResponseLoaded(it)
                    WRONG_CREDENTIALS -> responseCallback.onError(Constants.CREDENTIALS_ERROR)
                    else -> responseCallback.onError(Constants.UNKNOWN_ERROR)
                }
            }
        } catch (ex: Exception) {
            responseCallback.onError(Constants.CONNECTION_ERROR)
        }
    }

    private interface ResponseCallback {
        fun onResponseLoaded(response: Response)
        fun onError(error: String)
    }

    companion object {
        private var INSTANCE: PjatkAPI? = null
        private val WRONG_CREDENTIALS = 401
        private val OK = 200

        @JvmStatic
        fun getInstance(credentials: CredentialsManager.Credentials): StudentDataSource {
            return INSTANCE ?: PjatkAPI(credentials).apply { INSTANCE = this }
        }

        @JvmStatic
        fun destroyInstance() {
            INSTANCE = null
        }
    }
}